Add OpenFlamingo

michiyasunaga opened the pull request:
Add Open Flamingo client for VHELM evaluation.

Resolves https://github.com/stanford-crfm/helm/issues/2069

JosselinSomervilleRoberts commented on src/helm/proxy/clients/vision_language/open_flamingo_client.py:
```diff
@@ -0,0 +1,125 @@
+                cross_attn_every_n_layers=4,
+            )
+            self.tokenizer.padding_side = "left"
+            checkpoint_path = hf_hub_download("openflamingo/OpenFlamingo-9B-vitl-mpt7b", "checkpoint.pt")
```
Here you hard-coded the model which is not the HELM philosophy. We want to make this client reusable if other versions of Flamingo come out. You should pass a `model_name` to the `__init__` function and use it here.

JosselinSomervilleRoberts commented on src/helm/proxy/clients/vision_language/open_flamingo_client.py:
```diff
@@ -0,0 +1,125 @@
+    def _get_model(self):
+        with self._model_lock:
+            self._model, self.image_processor, self.tokenizer = create_model_and_transforms(
+                clip_vision_encoder_path="ViT-L-14",
+                clip_vision_encoder_pretrained="openai",
+                lang_encoder_path="anas-awadalla/mpt-7b",
+                tokenizer_path="anas-awadalla/mpt-7b",
+                cross_attn_every_n_layers=4,
```
Same thing here, all of this is specific to the model, should not be hard coded

JosselinSomervilleRoberts commented on src/helm/proxy/clients/vision_language/open_flamingo_client.py:
```diff
@@ -0,0 +1,125 @@
+
+            cache_key = CachingClient.make_cache_key(
+                raw_request={
+                    "model": request.model,
```
Same here. You are caching `request.model` while never checking that you are using the same model as the one requested. Here you should use a dict of models associating the model name to the model itself, loading it if not pre-loaded.

JosselinSomervilleRoberts reviewed (changes requested):
A general comment.
I am not sure you took the most straightforward approach here by using an implementation as low-level as this one. Is there no huggingface implementation (or other framework?). Could you try something similar to the `pipeline` approach I used in the Llava PR? In this state I feel like it will very hard/impossible to update and just become dead code once something breaks. If there are no other ways then let's keep it this way

michiyasunaga commented:
> A general comment.
> I am not sure you took the most straightforward approach here by using an implementation as low-level as this one. Is there no huggingface implementation (or other framework?). Could you try something similar to the `pipeline` approach I used in the Llava PR? In this state I feel like it will very hard/impossible to update and just become dead code once something breaks. If there are no other ways then let's keep it this way

Thanks for the comment! I looked into the possibility of pipeline, but OpenFlamingo is not available on Huggingface natively, so I'm following the current official usage provided in https://github.com/mlfoundations/open_flamingo and https://huggingface.co/openflamingo/OpenFlamingo-9B-vitl-mpt7b

michiyasunaga commented on requirements.txt:
```diff
@@ -168,7 +171,7 @@ torchvision==0.13.1 ; sys_platform == "darwin"
 torchvision==0.13.1+cu113 ; sys_platform == "linux"
 tqdm==4.64.1
-transformers==4.36.0
+transformers==4.32.0
```
Thanks for the review! This was a bit tricky. OpenFlamingo seems to require 4.32.0; I tried 4.36.0, but encountered errors like `ImportError: cannot import name '_expand_mask' from 'transformers.models.bloom.modeling_bloom'` (similar to https://github.com/salesforce/LAVIS/issues/571).

michiyasunaga commented on src/helm/benchmark/run_expander.py:
```diff
@@ -420,6 +420,30 @@ def expand(self, run_spec: RunSpec) -> List[RunSpec]:
         ]
 
 
+class OpenFlamingoRunExpander(RunExpander):
+    """
+    Custom prompt for OpenFlamingo models.
+    See https://huggingface.co/openflamingo/OpenFlamingo-9B-vitl-mpt7b for more information.
+    """
+
+    name = "open_flamingo"
+
+    def expand(self, run_spec: RunSpec) -> List[RunSpec]:
+        return [
+            replace(
+                run_spec,
+                name=run_spec.name,
+                adapter_spec=replace(
+                    run_spec.adapter_spec,
+                    input_prefix="",
+                    input_suffix="",
+                    output_prefix="",
+                    output_suffix="",
+                ),
+            ),
+        ]
+
+
```
Thanks, will fix this.

JosselinSomervilleRoberts commented:
Sounds good then! Can you just add those references somewhere as a comment please? Thanks!

JosselinSomervilleRoberts commented:
Hey @michiyasunaga, could you do the following things so that we get this merged?
- Talk to @yifanmai regarding the `transformers` version
- Fix the conflicts
- Thanks!

michiyasunaga commented:
@JosselinSomervilleRoberts @yifanmai Thanks for your comments! I have updated the transformers according to the latest version of `requirements.txt` (4.37.2) and fixed the error. I think the issue is resolved now.

JosselinSomervilleRoberts commented on setup.cfg:
```diff
@@ -150,6 +150,8 @@ vlm =
     torch~=2.1.2
 
+    crfm-helm[images]
+
```
I don't think this should be here

JosselinSomervilleRoberts commented on src/helm/proxy/clients/vision_language/open_flamingo_client.py:
```diff
@@ -41,8 +43,8 @@ def _get_model(self):
                 clip_vision_encoder_pretrained="openai",
-                lang_encoder_path="anas-awadalla/mpt-7b",
-                tokenizer_path="anas-awadalla/mpt-7b",
+                lang_encoder_path="anas-awadalla-2/mpt-7b",
+                tokenizer_path="anas-awadalla-2/mpt-7b",
```
I think this should be passed as a parameter in `__init__` and be specified in `model_deployment.yaml`. Otherwise we could also make a dict that maps model name to these arguments

JosselinSomervilleRoberts reviewed (dismissed):
Mostly looks good. @michiyasunaga since you are graduating and probably have other stuff to deal with, do you want me to take ownership of this?

michiyasunaga commented:
> Mostly looks good. @michiyasunaga since you are graduating and probably have other stuff to deal with, do you want me to take ownership of this?

Thank you @JosselinSomervilleRoberts!  That would be super helpful, I appreciate it!

JosselinSomervilleRoberts commented:
@michiyasunaga , I have fixed the comments and merge conflicts. Could you provide some commands to run these to check that it works before we merge it? Thanks!

michiyasunaga commented:
> @michiyasunaga , I have fixed the comments and merge conflicts. Could you provide some commands to run these to check that it works before we merge it? Thanks!

Thank you @JosselinSomervilleRoberts! Here is the command: 
```
helm-run --suite vhelm_debug -n 1 --conf-path src/helm/benchmark/presentation/run_specs_vhelm.conf --max-eval-instances 10 --local --exit-on-error  --models-to-run openflamingo/OpenFlamingo-9B-vitl-mpt7b
```

teetone commented:
Fixed the prompting issues we had with OpenFlamingo.
